client/src/main/java/io/pravega/client/stream/impl/EventStreamReaderImpl.java

```diff
@@ -258,10 +259,12 @@ private PositionInternal getCurrentPosition() {
     private String updateGroupStateIfNeeded() throws ReaderNotInReaderGroupException {
         PositionInternal position = null;
         if (atCheckpoint != null) {
+            // process the checkpoint we're at
             position = refreshAndGetPosition();
             groupState.checkpoint(atCheckpoint, position);
             log.info("Reader {} completed checkpoint {}", groupState.getReaderId(), atCheckpoint);
             releaseSegmentsIfNeeded(position);
+            groupState.updateTruncationStreamCutIfNeeded();
```

Comment:
are all readers reporting this streamcut? what happens if this fails 
1. because of intermittent issues
2. because the reader was reporting a stale value and a newer value was already published to controller

Reply:
Every reader would invoke that method, but it executes only of this condition is true:
```
if (state.getConfig().getRetentionConfig().equals(ReaderGroupConfig.RetentionConfig.CBR_LAST_CHECKPOINT)
        && !state.getCheckpointState().isLastCheckpointPublished()) {
```
isLastCheckpointPublished() is true by default and is changed to `false` only by the last reader completing a checkpoint :
https://github.com/pravega/pravega/blob/76e3a5be72635262f95b48d85880fe27954f4e12/client/src/main/java/io/pravega/client/stream/impl/CheckpointState.java#L154

So ideally this would only be executed by the lastReader completing the checkpoint and immediately after the streamcut is published to controller, we change the value of this variable back to "true" in state synchronizer, so no other Reader would attempt to publish again. If the streamcut publish from this reader fails, the variable value remains at false and so another Reader would publish the Stream-Cut.

Reply:
> because the reader was reporting a stale value and a newer value was already published to controller.

With regular fetchUpdates() being done I don't see the possibility of a reader trying to publish a stale checkpoint value and eventhat does happen the update would fail, based on streamcut validation checks added on Controller.